### PR TITLE
Fix referrer protections for frame redirects.

### PR DESCRIPTION
**Pull request by iefremov**
Fix https://github.com/brave/internal/issues/679
Fix https://github.com/brave/brave-browser/issues/7933

Resolves 

## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-browser/issues) for my issue if one did not already exist.
- [x] Used Github [auto-closing keywords](https://help.github.com/articles/closing-issues-via-commit-messages/) in the commit message.
- [x] Added/updated tests for this change (for new code or code which already has tests).
- Verified that these changes build without errors on
  - [ ] Android
  - [ ] iOS
  - [ ] Linux
  - [x] macOS
  - [ ] Windows
- Verified that these changes pass automated tests (unit, browser, security tests) on
  - [ ] iOS
  - [ ] Linux
  - [ ] macOS
  - [ ] Windows
- [ ] Verified that all lint errors/warnings are resolved (`npm run lint`, `npm run gn_check`)
- [x] Ran `git rebase master` (if needed).
- [x] Ran `git rebase -i` to squash commits (if needed).
- [x] Tagged reviewers and labelled the pull request as needed.
- [ ] Request a security/privacy review as needed.
- [ ] Add appropriate QA labels (QA/Yes or QA/No) to include the closed issue in milestone
- [ ] Public documentation has been updated as necessary. For instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-compatibility-issues-with-tracking-protection
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide

## Reviewer Checklist:

- [ ] New files have MPL-2.0 license header.
- [ ] Request a security/privacy review as needed.
- [ ] Adequate test coverage exists to prevent regressions 
- [ ] Verify test plan is specified in PR before merging to source

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on.
- [ ] All relevant documentation has been updated.

### CODE REVIEW

File: components/content_settings/renderer/brave_content_settings_agent_impl_browsertest.cc [hunk 1]
```diff
@@ -748,20 +746,18 @@ IN_PROC_BROWSER_TEST_F(BraveContentSettingsAgentImplBrowserTest,
...
-            cross_site_url().GetOrigin().spec());
-  EXPECT_EQ(GetLastReferrer(redirect_to_cross_site_url()),
-            cross_site_url().GetOrigin().spec());
+            url().GetOrigin().spec());
+  EXPECT_EQ(GetLastReferrer(cross_site_url()), url().GetOrigin().spec());
 
   // Cross-site navigations get no referrer.
   RedirectToPageWithLink(redirect_to_cross_site_url(), cross_site_url());
   EXPECT_EQ(ExecScriptGetStr(kReferrerScript, contents()), "");
-  EXPECT_EQ(GetLastReferrer(redirect_to_cross_site_url()), "");
```
Comment (fmarier) [inline components/content_settings/renderer/brave_content_settings_agent_impl_browsertest.cc:764]: Same as my previous comment, we should be able to check that the referrer of the intermediate steps are also correctly set.

**Review by fmarier — COMMENTED** [review]
I assume you've tested these changes using <https://referrer.fmarier.org/redirects.html>?

**Comment by iefremov** [comment]
>I assume you've tested these changes using https://referrer.fmarier.org/redirects.html?

sure thing

File: browser/net/brave_site_hacks_network_delegate_helper.cc [hunk 2]
```diff
@@ -147,7 +147,19 @@ int OnBeforeStartTransaction_SiteHacksWork(
...
     }
   }
+
+  // Special case for handling top-level redirects. There is no other way to
+  // normally change referrer in net::URLRequest during redirects.
+  if (!ctx->allow_referrers && ctx->allow_brave_shields &&
+      ctx->redirect_source.is_valid() &&
+      ctx->resource_type == blink::mojom::ResourceType::kMainFrame &&
+      brave_shields::ShouldCleanReferrerForTopLevelNavigation(
+          ctx->method, ctx->referrer, ctx->request_url)) {
```
[the remark this answers is not in the document]
Comment (iefremov) [reply browser/net/brave_site_hacks_network_delegate_helper.cc]: changed

File: components/content_settings/renderer/brave_content_settings_agent_impl_browsertest.cc (hunk 1 above)
Comment (iefremov) [reply components/content_settings/renderer/brave_content_settings_agent_impl_browsertest.cc:764]: added

File: components/content_settings/renderer/brave_content_settings_agent_impl_browsertest.cc [hunk 3]
```diff
@@ -650,20 +650,18 @@ IN_PROC_BROWSER_TEST_F(BraveContentSettingsAgentImplBrowserTest,
...
```
[the remark this answers is not in the document]
Comment (iefremov) [reply components/content_settings/renderer/brave_content_settings_agent_impl_browsertest.cc:666]: added (though it seems it should be `link_url()`

File: components/content_settings/renderer/brave_content_settings_agent_impl_browsertest.cc [hunk 4]
```diff
@@ -632,7 +632,7 @@ IN_PROC_BROWSER_TEST_F(BraveContentSettingsAgentImplBrowserTest,
 
 // TODO(iefremov): https://github.com/brave/brave-browser/issues/7933
```
[the remark this answers is not in the document]
Comment (iefremov) [reply components/content_settings/renderer/brave_content_settings_agent_impl_browsertest.cc]: fixed

File: components/content_settings/renderer/brave_content_settings_agent_impl_browsertest.cc [hunk 5]
```diff
@@ -734,7 +732,7 @@ IN_PROC_BROWSER_TEST_F(BraveContentSettingsAgentImplBrowserTest,
...
```
[the remark this answers is not in the document]
Comment (iefremov) [reply components/content_settings/renderer/brave_content_settings_agent_impl_browsertest.cc]: fixed

File: patches/net-url_request-url_request.cc.patch [hunk 6]
```diff
@@ -0,0 +1,12 @@
+diff --git a/net/url_request/url_request.cc b/net/url_request/url_request.cc
+index d96b34201073c6b90056b56730edcc5a83911044..1c533ab60234b244e0959b5081a53f991a79dd84 100644
+--- a/net/url_request/url_request.cc
++++ b/net/url_request/url_request.cc
+@@ -926,6 +926,7 @@ void URLRequest::Redirect(
+   method_ = redirect_info.new_method;
+   referrer_ = redirect_info.new_referrer;
+   referrer_policy_ = redirect_info.new_referrer_policy;
++  BRAVE_REDIRECT_CLEARREFERRER
```
Comment (iefremov) [inline patches/net-url_request-url_request.cc.patch:9]: @bridiver There is no way to update referrer in URLRequest (which actually sends this header to the network) via existing `URLLoader`/`URLLoaderClient` APIs , so I had to patch. An alternative was to use "extra" headers in `BraveProxyingURLLoaderFactory`, but earlier we decided to ban this approach due to potential negative performance implications.

File: components/content_settings/renderer/brave_content_settings_agent_impl_browsertest.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (fmarier) [reply components/content_settings/renderer/brave_content_settings_agent_impl_browsertest.cc:666]: Indeed, it should be `link_url()`.

**Review by fmarier — APPROVED** [review]

File: chromium_src/net/url_request/redirect_util.cc [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+    const base::Optional<net::HttpRequestHeaders>& modified_headers,
+    HttpRequestHeaders* request_headers,
+    bool* should_clear_upload) {
+  UpdateHttpRequest_ChromiumImpl(original_url,
+                                           original_method,
+                                           redirect_info,
+                                           removed_headers,
+                                           modified_headers,
+                                           request_headers,
+                                           should_clear_upload);
```
Comment (mkarolin) [inline chromium_src/net/url_request/redirect_util.cc:28]: Formatting looks a bit odd here.

**Review by bridiver — APPROVED** [review]
patches/chromium_src look ok

**Pull request merged**